contracts/bridge/contracts/common/fee-market/FeeMarket.sol

```diff
@@ -0,0 +1,416 @@
+        // No duplicate relayer allowed.
+        require(relayers[cur] == address(0), "!new");
+        // Prev relayer must in the list.
+        require(next != address(0), "!next");
+        // PrevFee <= MakerFee <= NextFee
+        require(fee >= feeOf[prev], "!>=");
+        require(fee <= feeOf[next], "!<=");
+        relayers[cur] = next;
+        relayers[prev] = cur;
+        feeOf[cur] = fee;
+        relayer_count++;
+        emit AddRelayer(prev, cur, fee);
+    }
+
+    // remove the relayer from the fee-market
+    function remove_relayer(address prev) public {
+        _remove_relayer(prev, msg.sender);
+    }
+
+    function prune_relayer(address prev, address cur) public {
+        if (lockedOf[cur] == 0 && balanceOf[cur] < COLLATERAL_PERORDER) {
+            _remove_relayer(prev, cur);
+        }
+    }
+
+    // move your position in the fee-market orderbook
+    function move_relayer(address old_prev, address new_prev, uint new_fee) public {
+        remove_relayer(old_prev);
+        add_relayer(new_prev, new_fee);
+    }
+
+    // Assign new message encoded key to top N relayers in fee-market
+    function assign(uint256 key) public override payable onlyOutBound returns (bool) {
+        //select top N relayers
+        address[] memory top_relayers = _get_and_prune_top_relayers();
+        address last = top_relayers[top_relayers.length - 1];
+        require(msg.value == feeOf[last], "!fee");
+        for (uint slot = 0; slot < top_relayers.length; slot++) {
+            address r = top_relayers[slot];
+            require(isRelayer(r), "!relayer");
+            _lock(r, COLLATERAL_PERORDER);
+            assigned_relayers[key][slot] = r;
+        }
+        // record the assigned time
+        orderOf[key] = block.timestamp;
+        emit OrderAssgigned(key, block.timestamp);
+        return true;
+    }
+
+    // Settle delivered messages and reward/slash relayers
+    function settle(DeliveredRelayer[] calldata delivery_relayers, address confirm_relayer) external override onlyOutBound returns (bool) {
+        _pay_relayers_rewards(delivery_relayers, confirm_relayer);
+        return true;
+    }
+
+    function _get_and_prune_top_relayers() internal returns (address[] memory) {
+        require(ASSIGNED_RELAYERS_NUMBER <= relayer_count, "!count");
+        address[] memory array = new address[](ASSIGNED_RELAYERS_NUMBER);
+        uint index = 0;
+        address prev = SENTINEL_HEAD;
+        address cur = relayers[SENTINEL_HEAD];
+        while (cur != SENTINEL_TAIL && index < ASSIGNED_RELAYERS_NUMBER) {
+            if (balanceOf[cur] >= COLLATERAL_PERORDER) {
+                array[index] = cur;
+                index++;
+            } else {
+                prune_relayer(prev, cur);
+            }
+            prev = cur;
+            cur = relayers[cur];
+        }
+        require(index == ASSIGNED_RELAYERS_NUMBER, "!assigned");
+        return array;
+    }
```

Comment:
simplify implementation and remove `prune_relayer`.

`balanceOf[cur]` should be sorted.

Reply:
`feeOf[src]` is sorted, not `balanceOf[cur]`